plugins/out_http/http.c

```diff
@@ -328,13 +366,167 @@ static int http_gelf(struct flb_out_http *ctx,
         s = tmp;
     }
 
-    ret = http_post(ctx, s, flb_sds_len(s), tag, tag_len);
+    ret = http_post(ctx, s, flb_sds_len(s), tag, tag_len, NULL);
     flb_sds_destroy(s);
     msgpack_unpacked_destroy(&result);
 
     return ret;
 }
 
+static char **extract_headers(msgpack_object *obj) {
+    size_t i;
+    char **headers = NULL;
+    size_t str_count;
+    msgpack_object_map map;
+    msgpack_object_str k;
+    msgpack_object_str v;
+
+    if (obj->type != MSGPACK_OBJECT_MAP) {
+        goto err;
+    }
+
+    map = obj->via.map;
+    str_count = map.size * 2 + 1;
+    headers = flb_calloc(str_count, sizeof *headers);
+
+    if (!headers) {
+        goto err;
+    }
+
+    for (i = 0; i < map.size; i++) {
+        if (map.ptr[i].key.type != MSGPACK_OBJECT_STR ||
+            map.ptr[i].val.type != MSGPACK_OBJECT_STR) {
+            continue;
+        }
+
+        k = map.ptr[i].key.via.str;
+        v = map.ptr[i].val.via.str;
+
+        headers[i * 2] = strndup(k.ptr, k.size);
+
+        if (!headers[i]) {
+            goto err;
+        }
+
+        headers[i * 2 + 1] = strndup(v.ptr, v.size);
+
+        if (!headers[i]) {
+            goto err;
+        }
+    }
+
+    return headers;
+
+err:
+    if (headers) {
+        for (i = 0; i < str_count; i++) {
+            if (headers[i]) {
+                flb_free(headers[i]);
+            }
+        }
+        flb_free(headers);
+    }
+    return NULL;
+}
+static int post_all_requests(struct flb_out_http *ctx,
+                             const char *data, size_t size,
+                             flb_sds_t body_key,
+                             flb_sds_t headers_key,
+                             struct flb_event_chunk *event_chunk)
+{
+    struct flb_time t;
+    msgpack_unpacked result;
+    msgpack_object root;
+    msgpack_object map;
+    msgpack_object *obj;
+    msgpack_object *k;
+    msgpack_object *v;
+    const char *body;
+    size_t body_size;
+    bool body_found;
+    bool headers_found;
+    char **headers;
+    size_t len;
+    size_t key_count;
+    size_t i;
+    size_t off = 0;
+    size_t record_count = 0;
+    int ret = 0;
+
+    msgpack_unpacked_init(&result);
+
+    while (msgpack_unpack_next(&result, data, size, &off) == MSGPACK_UNPACK_SUCCESS) {
+        headers = NULL;
+        body_found = false;
+        headers_found = false;
+        root = result.data;
+
+        if (root.type != MSGPACK_OBJECT_ARRAY) {
+            ret = -1;
+            break;
+        }
+
+        if (root.via.array.size != 2) {
+            ret = -1;
+            break;
+        }
+
+        flb_time_pop_from_msgpack(&t, &result, &obj);
+
+        map = root.via.array.ptr[1];
+        if (map.type != MSGPACK_OBJECT_MAP) {
+            ret = -1;
+            break;
+        }
+        key_count = map.via.map.size;
+
+        for (i = 0; i < key_count; i++) {
```

Comment:
I think this logic must be replaced by the record accessor functionality. Record accessor allows to specify a pattern and given a msgpack map it will extract the value. An example can be seen here:

- https://github.com/fluent/fluent-bit/blob/master/plugins/out_splunk/splunk_conf.c#L55

the pattern must be specified by the user like `$[arr1][other][the_real_key]` 

then on runtime, the code performs the lookup with:

- https://github.com/fluent/fluent-bit/blob/master/plugins/out_splunk/splunk.c#L71-L72

> note that this functionality allows the user to include the $tag as part of the final value